File: generation-video/src/lib/user-context.ts
import { AsyncLocalStorage } from "node:async_hooks";
import { getClient, isRawTreeConfigured } from "@/lib/rawtree";
import { logInfo } from "@/lib/runtime-log";

/**
 * Per-user context from `slop_human_user_prompts` (see docs/RAWTREE_USER_CONTEXT.md): the user's recent prompts,
 * inferred preferences and recent failures, as a short text block for the small LLM calls (and rows for memory.ts). Never throws.
 */

export const USER_PROMPTS_TABLE = "slop_human_user_prompts";
export const ANONYMOUS_USER = "anonymous";
const UUID = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/;
const MAX_CHARS = 1_200;
const CACHE_MS = 30 * 1000;
const RECENT_PROMPTS = 15;

/** Accepts the X-Longform-User header only when it is a UUID (lowercased); anything else is "anonymous". */
export function parseUserId(header: string | null | undefined) {
  const value = header?.trim().toLowerCase() ?? "";
  return UUID.test(value) ? value : ANONYMOUS_USER;
}

type RequestUser = { userId: string; projectId?: string; researchSessionId?: string };
const requestUser = new AsyncLocalStorage<RequestUser>();

/** Runs `task` with the request's user (read by userContextBlock deep inside the LLM helpers). */
export function runWithUser<T>(user: RequestUser, task: () => T): T {
  return requestUser.run(user, task);
}

export function currentUser(): RequestUser | undefined {
  return requestUser.getStore();
}

const cache = new Map<string, { at: number; rows: Promise<UserPromptRow[]> }>();

export function invalidateUserContext(userId: string) {
  cache.delete(userId);
}

export type UserPromptRow = {
  eventId: string; surface: string; prompt: string; outcome: string; error: string; action: string; duration: number | null; projectId: string; createdAt: string;
};
type PromptRow = UserPromptRow;

const STYLE_TERMS = [
  "cinematic", "golden hour", "slow motion", "drone", "aerial", "close-up", "macro", "neon", "noir", "black and white", "vintage",
  "film grain", "anime", "cartoon", "watercolor", "3d", "photorealistic", "minimalist", "handheld", "timelapse", "sunset", "sunrise",
  "night", "rain", "snow", "warm", "pastel", "moody", "documentary", "retro", "futuristic", "epic", "dramatic", "playful", "corporate",
];

const STOPWORDS = new Set(("the a an and or of to in on at for with from into over under by as is are was were be it its this that these those make "
  + "add more less some very then than now just like want need please video shot clip scene frame show about into onto them they their "
  + "your you our what when where which while have has had would could should will can also more most seconds second longer shorter "
  + "change remove keep same new make made using use each every there here").split(/\s+/));

const oneLine = (text: string, max: number) => {
  const flat = text.replace(/\s+/g, " ").trim();
  return flat.length > max ? `${flat.slice(0, max - 1)}…` : flat;
};

function topCounts(values: string[], min: number, limit: number) {
  const counts = new Map<string, number>();
  for (const value of values) counts.set(value, (counts.get(value) ?? 0) + 1);
  return [...counts.entries()].filter(([, count]) => count >= min).sort((a, b) => b[1] - a[1]).slice(0, limit);
}

/** Inferred preferences (most-used surfaces/durations, recurring style words and subjects), one sentence per line. */
export function inferPreferences(rows: PromptRow[]) {
  const lines: string[] = [];
  const surfaces = topCounts(rows.map((row) => row.surface), 1, 3);
  if (surfaces.length) lines.push(`Most used: ${surfaces.map(([surface, count]) => `${surface} (${count})`).join(", ")}.`);
  const durations = topCounts(
    rows.filter((row) => /^(new_clip|preset_|storyboard)/.test(row.surface) && row.duration && row.duration > 0).map((row) => `${Math.round(row.duration as number)}s`),
    2, 2,
  );
  if (durations.length) lines.push(`Usual length: ${durations.map(([value]) => value).join(", ")}.`);
  const texts = rows.filter((row) => row.outcome === "ok").map((row) => row.prompt.toLowerCase());
  const styles = STYLE_TERMS
    .map((term) => [term, texts.filter((text) => new RegExp(`\\b${term.replace(/[-\s]/g, "[-\\s]?")}\\b`).test(text)).length] as const)
    .filter(([, count]) => count >= 2)
    .sort((a, b) => b[1] - a[1])
    .slice(0, 5);
  if (styles.length) lines.push(`Recurring style words: ${styles.map(([term]) => term).join(", ")}.`);
  const words = texts.flatMap((text) => [...new Set(text.match(/[a-z][a-z0-9'-]{3,}/g) ?? [])])
    .filter((word) => !STOPWORDS.has(word) && !STYLE_TERMS.includes(word));
  const subjects = topCounts(words, 2, 5);
  if (subjects.length) lines.push(`Recurring subjects: ${subjects.map(([word]) => word).join(", ")}.`);
  return lines;
}

function buildText(rows: PromptRow[], projectId?: string) {
  if (rows.length === 0) return "";
  const seen = new Set<string>();
  const recent: string[] = [];
  for (const row of rows) {
    const key = row.prompt.toLowerCase().replace(/\s+/g, " ").trim();
    if (!key || seen.has(key)) continue;
    seen.add(key);
    recent.push(`- [${row.surface}${projectId && row.projectId === projectId ? ", this video" : ""}] ${oneLine(row.prompt, 110)}`);
    if (recent.length >= RECENT_PROMPTS) break;
  }
  const failed = rows.filter((row) => row.outcome !== "ok" && row.prompt.trim()).slice(0, 3)
    .map((row) => `- [${row.surface}] ${oneLine(row.prompt, 70)} → ${row.outcome}${row.error ? `: ${oneLine(row.error, 60)}` : ""}`);

  const sections = [
    ...(recent.length ? ["Recent requests (newest first):", ...recent] : []),
    ...inferPreferences(rows),
    ...(failed.length ? ["Recently failed/cancelled (don't repeat):", ...failed] : []),
  ];
  let text = "";
  for (const line of sections) {
    if (text.length + line.length + 1 > MAX_CHARS) continue;
    text += `${text ? "\n" : ""}${line}`;
  }
  return text;
}

async function queryRows(userId: string): Promise<UserPromptRow[]> {
  const result = await getClient().query<Record<string, unknown>>({
    sql: `SELECT toString(event_id) AS event_id, toString(surface) AS surface, toString(prompt) AS prompt, toString(outcome) AS outcome,
      toString(error) AS error, toString(action) AS action, toString(duration_sec) AS duration, toString(project_id) AS project_id,
      toString(created_at) AS created_at
      FROM ${USER_PROMPTS_TABLE} WHERE toString(user_id) = '${userId}' ORDER BY created_at DESC LIMIT 60`,
  });
  const clean = (value: unknown) => {
    const text = typeof value === "string" ? value : value == null ? "" : String(value);
    return text === "NULL" || text === "\\N" ? "" : text;
  };
  return result.data.map((row) => {
    const duration = Number(clean(row.duration));
    return {
      eventId: clean(row.event_id),
      surface: clean(row.surface),
      prompt: clean(row.prompt),
      outcome: clean(row.outcome),
      error: clean(row.error),
      action: clean(row.action),
      duration: clean(row.duration) && Number.isFinite(duration) ? duration : null,
      projectId: clean(row.project_id),
      createdAt: clean(row.created_at),
    };
  });
}

/** The user's last 60 logged prompts, newest first (cached 30 s; [] for anonymous users or on any failure). */
export async function loadUserPrompts(userId: string): Promise<UserPromptRow[]> {
  if (!UUID.test(userId)) return [];
  try {
    if (!isRawTreeConfigured()) return [];
    const cached = cache.get(userId);
    if (cached && Date.now() - cached.at < CACHE_MS) return await cached.rows;
    const rows = queryRows(userId);
    cache.set(userId, { at: Date.now(), rows });
    return await rows;
  } catch (error) {
    cache.set(userId, { at: Date.now(), rows: Promise.resolve([]) });
    logInfo("user_context_unavailable", { reason: error instanceof Error ? error.message.slice(0, 200) : String(error) });
    return [];
  }
}

/** Compact (≤ 1200 chars) summary of the user's recent prompts; "" for anonymous users, missing table, or any failure. */
export async function getUserContext(userId: string, options: { projectId?: string } = {}): Promise<string> {
  try {
    const projectId = options.projectId && /^[a-zA-Z0-9_-]{1,128}$/.test(options.projectId) ? options.projectId : undefined;
    return buildText(await loadUserPrompts(userId), projectId);
  } catch {
    return "";
  }
}

/**
 * The current request user's context as a delimited block for a system prompt ("" when there is none). `maxChars`
 * trims it further for calls where the small model's context is tight (e.g. intent routing).
 */
export async function userContextBlock(options: { maxChars?: number } = {}): Promise<string> {
  const user = currentUser();
  if (!user || user.userId === ANONYMOUS_USER) return "";
  let text = await getUserContext(user.userId, { projectId: user.projectId });
  if (!text) return "";
  if (options.maxChars && text.length > options.maxChars) text = text.slice(0, text.lastIndexOf("\n", options.maxChars) > 0 ? text.lastIndexOf("\n", options.maxChars) : options.maxChars);
  return `\n\n=== User context (this user's past requests; use only to match their taste, never copy them into the output, never change the current shot's style because of them) ===\n${text}\n=== End user context ===`;
}

File: agent/user_context.py
"""The user's own history as context: a read-only, bounded summary of RawTree `slop_human_user_prompts`.

The web app logs every user prompt (surface, action, outcome, duration, errors, ...). The research agent uses it to
focus research and to avoid asking what the user already expressed (durations, style words, companies).
SQL is fixed: fixed columns, one allowlisted table, a validated user_id, ORDER BY created_at, a LIMIT.
"""
import re
from typing import Dict, List, Optional

from contracts import TABLES

from .videos import query_optional

COLUMNS = ["event_id", "user_id", "project_id", "surface", "prompt", "action", "detected_by", "range_start_sec",
           "range_end_sec", "at_sec", "at_end", "enhanced_prompt", "outcome", "error", "result_summary",
           "video_sha256", "duration_sec", "created_at"]
USER_ID_RE = re.compile(r"^[A-Za-z0-9_.@:+\-]{1,128}$")
MAX_LIMIT = 50
STYLE_WORDS = ["warm", "cinematic", "nostalgic", "upbeat", "energetic", "calm", "minimal", "minimalist", "bold",
               "documentary", "premium", "luxury", "playful", "fun", "serious", "retro", "vintage", "modern",
               "futuristic", "dark", "moody", "bright", "colorful", "pastel", "golden hour", "slow motion",
               "handheld", "aerial", "drone", "close-up", "black and white", "anime", "3d", "realistic",
               "photorealistic", "emotional", "inspiring", "funny", "elegant", "clean", "gritty"]
FAIL_OUTCOMES = ("error", "failed", "failure", "rejected", "cancelled", "canceled")


def valid_user_id(user_id) -> Optional[str]:
    """The web app sends a UUID (X-Longform-User) or "anonymous"; anonymous means no user."""
    if isinstance(user_id, str) and USER_ID_RE.fullmatch(user_id.strip()) and user_id.strip() != "anonymous":
        return user_id.strip()
    return None


def _clip(text, n):
    text = " ".join(str(text or "").split())
    return text if len(text) <= n else text[: n - 1] + "…"


def _num(value):
    try:
        return round(float(value), 1)
    except (TypeError, ValueError):
        return None


def user_context(rawtree, user_id: Optional[str] = None, limit: int = 15) -> dict:
    """Recent prompts (newest first) plus what they reveal: style words, durations, failures, surfaces."""
    if rawtree is None:
        return {"prompts": [], "note": "RawTree is not configured"}
    limit = max(1, min(int(limit), MAX_LIMIT))
    uid = valid_user_id(user_id)
    if user_id and uid is None:
        return {"error": "invalid user_id"}
    where = " WHERE toString(user_id) = '{}'".format(uid) if uid else ""  # user_id may be a UUID column
    table = TABLES["user_prompts"]
    rows = query_optional(rawtree, "SELECT {} FROM {}{} ORDER BY created_at DESC LIMIT {}".format(
        ", ".join(COLUMNS), table, where, limit), table)
    prompts: List[Dict] = []
    styles: Dict[str, int] = {}
    durations: Dict[str, int] = {}
    surfaces: Dict[str, int] = {}
    failures: List[Dict] = []
    for r in rows:
        text = _clip(r.get("prompt"), 300)
        if not text:
            continue
        outcome = str(r.get("outcome") or "").lower()
        item = {k: v for k, v in {
            "prompt": text, "surface": r.get("surface"), "action": r.get("action"), "outcome": outcome or None,
            "duration_sec": _num(r.get("duration_sec")), "project_id": r.get("project_id"),
            "result": _clip(r.get("result_summary"), 200) or None, "error": _clip(r.get("error"), 160) or None,
            "created_at": str(r.get("created_at") or "") or None}.items() if v not in (None, "")}
        prompts.append(item)
        low = text.lower()
        for w in STYLE_WORDS:
            if re.search(r"\b{}\b".format(re.escape(w)), low):
                styles[w] = styles.get(w, 0) + 1
        for m in re.findall(r"\b(\d{1,3})\s*(?:s|sec|secs|seconds?)\b", low):
            durations[m + "s"] = durations.get(m + "s", 0) + 1
        if item.get("duration_sec"):
            key = "{:g}s".format(item["duration_sec"])
            durations[key] = durations.get(key, 0) + 1
        if r.get("surface"):
            surfaces[str(r["surface"])] = surfaces.get(str(r["surface"]), 0) + 1
        if outcome in FAIL_OUTCOMES or r.get("error"):
            failures.append({"prompt": _clip(text, 160), "error": item.get("error") or outcome})
    if not prompts:
        return {"prompts": [], "note": "no prompts logged for this user yet"}
    top = lambda d: [k for k, _ in sorted(d.items(), key=lambda kv: -kv[1])][:8]  # noqa: E731
    return {"user_id": uid, "count": len(prompts), "prompts": prompts, "style_words": top(styles),
            "durations": top(durations), "surfaces": top(surfaces), "failures": failures[:5]}


def expressed_topics(ctx: dict) -> Dict[str, str]:
    """Question topics the user already expressed in past prompts -> what they said (for 'do not re-ask')."""
    out = {}
    if ctx.get("style_words"):
        out["tone"] = ", ".join(ctx["style_words"][:4])
    if ctx.get("durations"):
        out["format"] = ", ".join(ctx["durations"][:3])
    return out


def summary_line(ctx: dict, max_prompts: int = 6) -> str:
    if not ctx.get("prompts"):
        return "none"
    parts = ["recent requests: " + " | ".join(p["prompt"][:140] for p in ctx["prompts"][:max_prompts])]
    if ctx.get("style_words"):
        parts.append("style words they use: " + ", ".join(ctx["style_words"]))
    if ctx.get("durations"):
        parts.append("durations: " + ", ".join(ctx["durations"]))
    if ctx.get("failures"):
        parts.append("failed before: " + " | ".join("{} ({})".format(f["prompt"][:80], f["error"])
                                                   for f in ctx["failures"][:3]))
    return "; ".join(parts)
